test/shopify_app/auth/token_exchange_test.rb

```diff
@@ -0,0 +1,175 @@
+# frozen_string_literal: true
+
+require "test_helper"
+
+class ShopifyApp::Auth::TokenExchangeTest < ActiveSupport::TestCase
+  OFFLINE_ACCESS_TOKEN_TYPE = ShopifyAPI::Auth::TokenExchange::RequestedTokenType::OFFLINE_ACCESS_TOKEN
+  ONLINE_ACCESS_TOKEN_TYPE = ShopifyAPI::Auth::TokenExchange::RequestedTokenType::ONLINE_ACCESS_TOKEN
+
+  def setup
+    ShopifyApp::SessionRepository.shop_storage = ShopifyApp::InMemoryShopSessionStore
+    ShopifyApp::SessionRepository.user_storage = nil
+
+    @shop = "my-shop.myshopify.com"
+    @user_id = 1
+    @session_token = build_jwt
+
+    @offline_session = build_offline_session
+    @online_session = build_online_session
+  end
+
+  test "#perform exchanges offline token then stores it when only shop session store is configured" do
+    ShopifyAPI::Auth::TokenExchange.expects(:exchange_token).with(
+      shop: @shop,
+      session_token: @session_token,
+      requested_token_type: OFFLINE_ACCESS_TOKEN_TYPE,
+    ).returns(@offline_session)
```

Comment:
Do we have to explicitly say we only expect one call here?

Reply:
Mocha's `expects` fails if not called [exactly once](https://mocha.jamesmead.org/Mocha/Mock.html#expects-instance_method), but there's also a `.once` method for readability which we can use to make it more explicit.